R/escape.R

```diff
@@ -4,21 +4,51 @@
 #' @param seuratObj A Seurat object.
 #' @param outputAssayName The name of the assay to store results
 #' @param doPlot If true, a FeaturePlot will be printed for each pathway
-#' @param geneSets Which gene sets should be fetched from MSigDB and calculated. Some recommendations in increasing computation time: H (hallmark, 50 gene sets), C8 (scRNASeq cell type markers, 830 gene sets), C2 (curated pathways, 6366 gene sets), GO:BP (GO biological processes, 7658). 
+#' @param msigdbGeneSets A vector containing gene set codes specifying which gene sets should be fetched from MSigDB and calculated. Some recommendations in increasing computation time: H (hallmark, 50 gene sets), C8 (scRNASeq cell type markers, 830 gene sets), C2 (curated pathways, 6366 gene sets), GO:BP (GO biological processes, 7658). 
+#' @param customGeneSets A (preferably named) list containing gene sets to be scored by escape. 
 #' @return The seurat object with results stored in an assay
 #' @export
-RunEscape <- function(seuratObj, outputAssayName = "escape.ssGSEA", doPlot = FALSE, geneSets = c("H")) {
+RunEscape <- function(seuratObj, outputAssayName = "escape.ssGSEA", doPlot = FALSE, msigdbGeneSets = c("H"), customGeneSets = NULL) {
+  #gene set vector to be populated by msigdb or custom gene sets
+  GS <- c()
   
-  ##GO:BP and other hierarchical gene sets require subcategories to be passed with them, so we should parse those individually and concatenate afterwards. I think each of these need to be individually supported, but I haven't confirmed. 
-  #check if all gene sets are non-hierarchical
-  if (all(geneSets %in% c("H", paste0("C", 1:8)))) {
-    GS <- escape::getGeneSets(library = geneSets)
-  } else if ("GO:BP" %in% geneSets) {
-    geneSets <- geneSets[geneSets != "GO:BP"]
-    GS_GO_BP <- escape::getGeneSets(library = "C5", subcategory = "BP")
-    GS <- c(escape::getGeneSets(library = geneSets), GS_GO_BP)
-  } else {
-    stop("geneSets is not one of: H, GO:BP",paste0(", C", 1:8), ". Please ensure your requested geneSet is listed, or contact members of the Bimber Lab to add a gene set.")
+  #ensure msigdbGeneSets is formatted properly to parse. 
+  if (!is.null(msigdbGeneSets) & !is.na(msigdbGeneSets) & !(length(msigdbGeneSets) == 0)) {
```

Comment:
I think it's better practice to make this use double ampersands (&&, not &), and to check (all(is.null())). the point is that if() cant handle vectors

Reply:
good point, thanks!